### ISSUE TITLE
Checkpoint best model instead of n most recent

### ISSUE THREAD

**Issue by ltiao** [data61/aboleth#159]
Find a clean way to checkpoint the current best model, according to some scalar metric on the hold-out validation set, instead of the default most recent criteria. See e.g.
- https://stackoverflow.com/questions/39252901/tensorflow-save-the-model-with-smallest-validation-error
- https://github.com/Hvass-Labs/TensorFlow-Tutorials/blob/master/04_Save_Restore.ipynb

**Comment by ltiao** [comment]
The right way would be to subclass 
[SessionRunHook](https://github.com/tensorflow/tensorflow/blob/v1.4.0/tensorflow/python/training/session_run_hook.py). Should look very similar to [SummarySaverHook](https://www.tensorflow.org/api_docs/python/tf/train/SummarySaverHook), except we save if the current model beats the current *best* model on hold-out test set performance, instead of saving every n steps or seconds.

**Comment by ltiao** [comment]
The minimal viable way would be simply to perform all this logic in the training loop.

**Comment by ltiao** [comment]
c.f. 96fa85f for latest commit in WIP branch on this issue ( @lmccalman)

**Comment by dsteinberg** [comment]
this has been done externally to this repo (see private landshark repo)
